pkg/model/mselection/heaviest.go

```diff
@@ -20,137 +19,230 @@ import (
 
-	// tips could be empty after a reset
-	if len(tips) == 0 {
-		return nil, ErrNoTipsAvailable
+	if tipsList.Len() == 0 {
+		return nil, 0, ErrNoTipsAvailable
 	}
 
-	lastTip := tips[len(tips)-1]
+	lastTip := tipsList.items[tipsList.Len()-1]
 
 	var best = struct {
-		tips  hornet.Hashes
+		tips  []*selectedTip
 		count uint
 	}{
-		tips:  hornet.Hashes{lastTip.hash},
+		tips: []*selectedTip{
+			{
+				item:  lastTip,
+				index: tipsList.Len() - 1,
+			}},
 		count: lastTip.refs.Count(),
 	}
 
 	// loop through all tips and find the one with the most referenced transactions
-	for _, tip := range tips {
+	for index, tip := range tipsList.items {
 		// when the context has been cancelled, return the current best with an error
 		select {
 		case <-ctx.Done():
-			return randomTip(best.tips), ctx.Err()
+			selected, err := randomTipFromTips(best.tips)
+			if err != nil {
+				return nil, 0, err
+			}
+			return selected, best.count, ctx.Err()
 		default:
 		}
 
 		c := tip.refs.Count()
 		if c > best.count {
-			best.tips = hornet.Hashes{tip.hash}
+			// tip with heavier branch found
+			best.tips = []*selectedTip{{
+				item:  tip,
+				index: index,
+			}}
 			best.count = c
 		} else if c == best.count {
-			best.tips = append(best.tips, tip.hash)
+			// add the tip to the slice of currently best tips
+			best.tips = append(best.tips, &selectedTip{
+				item:  tip,
+				index: index,
+			})
 		}
 	}
 
-	// TODO: is it really to select a random tip? Maybe prefer the older (or younger) tip instead
-	return randomTip(best.tips), nil
+	selected, err := randomTipFromTips(best.tips)
+	if err != nil {
+		return nil, 0, err
+	}
+	return selected, best.count, nil
 }
 
-// SelectTip selects a tip to be used for the next milestone.
-func (s *HeaviestSelector) SelectTip() (hornet.Hash, error) {
+// SelectTips tries to collect tips that confirm the most transactions in the future cone.
+// best tips are determined by counting the referenced transactions (heaviest branches) and by "removing" the
+// transactions of the referenced cone of the already choosen tips in the bitsets of the available tips.
+// only tips are considered that were present at the beginning of the SelectTips call,
+// to prevent attackers from creating heavier branches while we are searching the best tips.
+// "maxHeaviestBranchTipsPerCheckpoint" is the amount of tips that are collected if
+// the current best tip is not below "UnconfirmedTransactionsThreshold" before.
+// selecting at least one tip can be enforced, even if none of the heaviest branches matches the
+// "minHeaviestBranchUnconfirmedTransactionsThreshold" criteria with "enforceTips".
+// if at least one heaviest branch tip was found, "randomTipsPerCheckpoint" random tips are added
+// to add some additional randomness to prevent parasite chain attacks.
+func (s *HeaviestSelector) SelectTips(enforceTips bool) (hornet.Hashes, error) {
```

Comment:
The `SelectTips` is actual a greedy heuristic algorithm. It iteratively selects the "heaviest" tip, removes its past cone and then continues with the now heaviest tip until the required number of tips is reached. This will not be the optimal solution.
The heaviest pair selection actually selected the optimal two tips which can be considerably stronger than the first two tips returned by this.
I would still be very very much interested to see how this behaves. It would be my guess, that the the quality is actually very much similar but it would work like the current protocol with only one Milestone bundle which could be preferable.